packages/rum-core/test/common/fetch-patch.spec.js

```diff
@@ -49,20 +49,146 @@ describe('fetchPatch', function () {
+      })
+
+      it('should fetch correctly', function (done) {
+        var promise = window.fetch('/')
+        expect(promise).toBeDefined()
+        expect(typeof promise.then).toBe('function')
+        expect(events.map(e => e.event)).toEqual(['schedule'])
+        promise.then(function (resp) {
+          expect(resp).toBeDefined()
+          Promise.resolve().then(function () {
+            expect(events.map(e => e.event)).toEqual(['schedule', 'invoke'])
+            done()
+          })
+        })
+      })
+
+      it('should produce task events when fetch fails', function (done) {
+        var promise = window.fetch('http://localhost:54321/')
+        promise.catch(function (error) {
+          expect(error).toBeDefined()
+          Promise.resolve().then(function () {
+            expect(events.map(e => e.event)).toEqual(['schedule', 'invoke'])
+            done()
+          })
+        })
+      })
+
+      it('should set invoke task as aborted when fetch request aborts', function (done) {
+        const abortController = new AbortController()
+        var promise = window.fetch('/', { signal: abortController.signal })
+        abortController.abort()
+
+        promise.catch(function (resp) {
+          expect(resp).toBeDefined()
+          Promise.resolve().then(function () {
+            expect(events.map(e => e.event)).toEqual(['schedule', 'invoke'])
+            const invokeTask = events[1].task
+            expect(invokeTask.data.aborted).toBe(true)
+            done()
+          })
         })
       })
     })
 
+    if (window.ReadableStream) {
+      describe('When response can be read from a stream', () => {
+        function mockStreamReader(error) {
+          let resolveStream
+          let rejectStream
+          const streamFinished = new Promise((resolve, reject) => {
+            resolveStream = resolve
+            rejectStream = reject
+          })
+
+          spyOn(ReadableStream.prototype, 'getReader').and.callFake(() => {
+            let readCounter = 0
+
+            return {
+              read: () => {
+                if (error) {
+                  rejectStream()
+                  return Promise.reject(error)
+                }
+                if (readCounter > 0) {
+                  const response = Promise.resolve({
+                    value: undefined,
+                    done: true
+                  })
+
+                  resolveStream()
+
+                  return response
+                }
+
+                readCounter += 1
+                return Promise.resolve({ value: 'test', done: false })
+              }
+            }
+          })
+
+          return streamFinished
+        }
+
+        it('should produce task events when all data have been read', function (done) {
+          const streamFinished = mockStreamReader()
+
+          window.fetch('/')
+          streamFinished.then(() => {
+            Promise.resolve().then(function () {
```

Comment:
Do we need this microtask `Promise.resolve` here? I guess the stream already resolves after the microtask, which is what made me curious.

Reply:
We need that **microtask** there, because in the `getReader` fake we are resolving the dummy promise before the getReader one:

```
    if (readCounter > 0) {
      const response = Promise.resolve({
        value: undefined,
        done: true
      })

      resolveStream()

      return response
    }
```

The microtask makes sure the expect will find the proper data.

Another way of doing this is to resolve the dummy at the end:

```
    if (readCounter > 0) {
      const response = Promise.resolve({
        value: undefined,
        done: true
      })

      // a macrotask here
      setTimeout(resolveStream, 0) 

      return response
    }
```

With this we wouldn't need the `Promise.resolve` in the test. I'm not have a strong opinion about this. Both ways are okay to me (since other tests were already doing something similar as well). WDYT?